samples: nrf5340: extxip_smp_svr: Convert to sysbuild (#16513)
nrfconnect/sdk-nrf: doc/nrf/app_dev/bootloaders_dfu/mcuboot_nsib/qspi_xip_split_image.rst

.. _qspi_xip_split_image:

Using QSPI XIP split image
##########################

.. contents::
   :local:
   :depth: 2

The QSPI XIP split image feature lets you gain more flash storage space for applications by splitting application code into two parts:

* Code that runs on the internal flash memory.
* Code that runs on supported external flash memory over the Quad Serial Peripheral Interface (QSPI) using Execute in Place (XIP).

This feature is supported on nRF52840, nRF5340, and the SoCs of the nRF91 Series.

.. warning::
    On the nRF52840, do not relocate interrupts to the QSPI XIP flash.
    Doing so can lock up or brick the device by making the debug access port inaccessible.

The QSPI XIP split images are supported in :doc:`MCUboot <mcuboot:index-ncs>`, which allows for updating them over-the-air.

For the nRF5340 DK and Nordic Thingy:53, you can also check out the :ref:`SMP Server with external XIP <smp_svr_ext_xip>` sample, which demonstrates this feature.

Requirements
************

> Make sure you meet the following requirements to make use of this feature:

Author: Fixed

* Board based on nRF52840, nRF5340, or nRF91 Series SoCs, with file structure compatible with :ref:`Hardware model v2 (HWMv2) <hwmv1_to_v2_migration>`
* External QSPI flash chip with supported commands connected to QSPI pins
* QSPI flash chip in always-on mode (meaning, no DPM or low-power modes)
* :doc:`MCUboot configuration <mcuboot:design>` in the Swap using move mode (``MCUBOOT_SWAP_USING_MOVE``), the Upgrade only mode (``MCUBOOT_OVERWRITE_ONLY``), or in the direct-XIP mode
* :ref:`Static partition manager file <ug_pm_static>`
* Linker file with QSPI XIP flash offset and size
* :ref:`configuration_system_overview_sysbuild`

.. caution::

>     At present, MCUboot images cannot be linked together.

Author: Fixed

    This means that both parts of the firmware update must be loaded properly before an update is initiated.

>     If one part fails to update, the image on the device will be unbootable and will likely fault.

Author: yes

Network core image updates
**************************

> The QSPI XIP split image feature can be used and will retain support for network core image updates on nRF5340 devices if MCUboot is configured for the Swap using move mode or the Upgrade only mode.

Author: Fixed

Relocate the code to run from the QSPI flash inside the applications :file:`CMakeLists.txt` file using the ``zephyr_code_relocate`` function.
This can be used to relocate files and libraries.

.. tabs::

   .. group-tab:: Relocating files

      .. code-block:: cmake

         zephyr_code_relocate(FILES src/complex_sensor_calculcation_code.c LOCATION EXTFLASH_TEXT NOCOPY)
         zephyr_code_relocate(FILES src/complex_sensor_calculcation_code.c LOCATION RAM_DATA)

      .. note::
          This assumes that the application has a :file:`src/complex_sensor_calculcation_code.c`` file.

   .. group-tab:: Relocating libraries

      .. code-block:: cmake

         zephyr_code_relocate(LIBRARY subsys__mgmt__mcumgr__mgmt LOCATION EXTFLASH_TEXT NOCOPY)
         zephyr_code_relocate(LIBRARY subsys__mgmt__mcumgr__mgmt LOCATION RAM_DATA)

      .. note::
          The library name comes from the Zephyr code.

.. rst-class:: numbered-step

Set QSPI flash initialization priority
**************************************

Set the initialization priority of the QSPI flash so that it is configured before any code on the device runs, otherwise the application operation is going to be undefined.

> There is no in-built check for this setting, so the priority must be set manually based upon the code that has been relocated.

Reviewer: ```suggestion
There is no built-in check for this setting.
Therefore, you must manually set the priority based on the relocated code.
```

Author: Fixed

> For example, if you relocate the shell subsystem, the initialization priority of any of that subsystem must be higher than the priority of QSPI, and the value of :kconfig:option:`CONFIG_SHELL_BACKEND_SERIAL_INIT_PRIORITY` must be greater than :kconfig:option:`CONFIG_NORDIC_QSPI_NOR_INIT_PRIORITY`.

Reviewer: ```suggestion
For example, if you relocate the shell subsystem, its initialization priority must be higher than the priority of QSPI, and the value of :kconfig:option:`CONFIG_SHELL_BACKEND_SERIAL_INIT_PRIORITY` must be greater than :kconfig:option:`CONFIG_NORDIC_QSPI_NOR_INIT_PRIORITY`.
```

Author: Fixed

.. note::
    Not all libraries or subsystems can be relocated.
    Any code that runs early in the boot process or uses ``SYS_INIT()`` with a non-configurable init priority cannot be relocated.
    Library and subsystem code needs to be manually checked before relocating it.

Programming with the QSPI XIP split image
*****************************************

Programming of the application is supported using the :ref:`standard procedure <programming>`.
The standard procedure will program the firmware using the default nrfjprog configuration which, for QSPI, is PP4IO mode.

Programming using a different SPI mode
======================================

If you are using an a different SPI mode on the QSPI interface, such as DSPI, you must use a custom :file:`Qspi.ini` file.

> Here's an example for the Thingy:53, which only supports DSPI and PP:

Reviewer: ```suggestion
The following is an example for the Thingy:53, which supports DSPI and PP:
```

Author: Fixed

.. literalinclude:: ../../../../../samples/nrf5340/extxip_smp_svr/Qspi_thingy53.ini

To use this file when programming, add the following lines to the application's :file:`CMakeLists.txt` file before the ``find_package()`` line:

.. code-block:: cmake

    macro(app_set_runner_args)
      # Replace with the filename of your ini file
      board_runner_args(nrfjprog "--qspiini=${CMAKE_CURRENT_SOURCE_DIR}/Qspi_thingy53.ini")
    endmacro()

This will enable programming the target board successfully when using ``west flash``.

Firmware updates
****************

The build system will automatically generate the signed update files that can be loaded to a device to run the newer version of the application.
The :file:`dfu_application.zip` file also contains these files and you can use the `nRF Connect for Mobile`_ application (available for Android and iOS) to upload the firmware over Bluetooth® Low Energy.

File descriptions
*****************

> The following table lists files that are generated when building a QSPI XIP split image application.

Reviewer: ```suggestion
The following table lists the files generated when building a QSPI XIP split-image application.
```

Author: Fixed

The QSPI XIP split images are supported in :doc:`MCUboot <mcuboot:index-ncs>`, which allows for updating them over-the-air.

For the nRF5340 DK and Nordic Thingy:53, you can also check out the :ref:`SMP Server with external XIP <smp_svr_ext_xip>` sample, which demonstrates this feature.

Requirements
************

To use this feature, meet the following requirements:

* Board based on nRF52840, nRF5340, or nRF91 Series SoCs, with file structure compatible with :ref:`Hardware model v2 (HWMv2) <hwmv1_to_v2_migration>`
* External QSPI flash chip with supported commands connected to QSPI pins
* QSPI flash chip in always-on mode (meaning, no DPM or low-power modes)
* :doc:`MCUboot configuration <mcuboot:design>` in the Swap using move mode (``MCUBOOT_SWAP_USING_MOVE``), the Upgrade only mode (``MCUBOOT_OVERWRITE_ONLY``), or in the direct-XIP mode
* :ref:`Static partition manager file <ug_pm_static>`
* Linker file with QSPI XIP flash offset and size
* :ref:`configuration_system_overview_sysbuild`

.. caution::
    Currently, MCUboot images cannot be linked together.
    This means that both parts of the firmware update must be loaded properly before an update is initiated.

>     If one part fails to update, the image on the device will be unbootable and will likely fault.

Reviewer: ```suggestion
    If one part fails to update, the image on the device will be unbootable and that will likely brick the device.
```

Author: Fixed
